code/ecdsap256sha2/Hacl.Impl.ECDSA.P256SHA256.Signature.Agile.fst

```diff
@@ -0,0 +1,319 @@
+
+open Hacl.Impl.ECDSA.MM.Exponent
+open Hacl.Impl.ECDSA.MontgomeryMultiplication
+
+open Hacl.Impl.P256.Signature.Common
+
+module H = Spec.Agile.Hash
+module Def = Spec.Hash.Definitions
+
+open Spec.Hash.Definitions
+open Hacl.Hash.Definitions
+
+#set-options "--z3rlimit 100"
+
+inline_for_extraction noextract
+val ecdsa_signature_step12: alg: hash_alg {SHA2_256? alg \/ SHA2_384? alg \/ SHA2_512? alg (* and blake one day *)}
+  ->  mLen: size_t -> m: lbuffer uint8 mLen -> result: felem -> Stack unit
+  (requires fun h -> live h m /\ live h result )
+  (ensures fun h0 _ h1 -> modifies (loc result) h0 h1 /\
+    (
+      assert_norm (pow2 32 < pow2 61);
+      assert_norm (pow2 32 < pow2 125);
+      let hashM = (hashSpec alg) (as_seq h0 m) in 
+      let cutHashM = Lib.Sequence.sub hashM 0 32 in 
+      as_nat h1 result = nat_from_bytes_be cutHashM % prime_p256_order
+    )
+  )
+
+
+let ecdsa_signature_step12 alg mLen m result = 
+  assert_norm (pow2 32 < pow2 61);
+  assert_norm (pow2 32 < pow2 125);
+  push_frame(); 
+    let h0 = ST.get() in 
+  let sz: FStar.UInt32.t = hash_len alg in
+  let mHash = create sz (u8 0) in    
+  
+  begin
+  match alg with 
+    |SHA2_256 ->
+      hash_256 m mLen mHash
+    |SHA2_384 ->
+      hash_384 m mLen mHash
+    |SHA2_512 -> 
+      hash_512 m mLen mHash
+  end;
+  
+  let cutHash = sub mHash (size 0) (size 32) in 
+  toUint64ChangeEndian cutHash result;
+  
+  let h1 = ST.get() in 
+ 
+  reduction_prime_2prime_order result result;
+
+  lemma_core_0 result h1;
+  Spec.ECDSA.changeEndianLemma (uints_from_bytes_be #U64 #_ #4 (as_seq h1 cutHash));
+  uints_from_bytes_be_nat_lemma #U64 #_ #4 (as_seq h1 cutHash);
+
+  pop_frame()
+
+
+inline_for_extraction
+val ecdsa_signature_step45: x: felem -> k: lbuffer uint8 (size 32) -> tempBuffer: lbuffer uint64 (size 100) -> Stack uint64
+  (requires fun h -> 
+    live h x /\ live h k /\ live h tempBuffer /\ 
+    LowStar.Monotonic.Buffer.all_disjoint [loc tempBuffer; loc k; loc x]
+  )
+  (ensures fun h0 r h1 -> 
+    modifies (loc x |+| loc tempBuffer) h0 h1 /\ 
+    as_nat h1 x < prime_p256_order /\ 
+    (
+      let (rxN, ryN, rzN), _ = montgomery_ladder_spec (as_seq h0 k) ((0,0,0), basePoint) in 
+      let (xN, _, _) = _norm (rxN, ryN, rzN) in 
+      as_nat h1 x == xN % prime_p256_order /\ 
+      (
+  if as_nat h1 x = 0 then uint_v r == pow2 64 - 1 else uint_v r == 0
+      )
+    )
+  )
+
+let ecdsa_signature_step45 x k tempBuffer = 
+  push_frame();
+    let result = create (size 12) (u64 0) in 
+    let tempForNorm = sub tempBuffer (size 0) (size 88) in 
+    secretToPublicWithoutNorm result k tempBuffer; 
+    normX result x tempForNorm;
+    reduction_prime_2prime_order x x;
+  pop_frame();
+    isZero_uint64_CT x
+
+
+val lemma_power_step6: kInv: nat -> Lemma 
+  (Spec.ECDSA.exponent_spec (fromDomain_ kInv) == toDomain_ (pow kInv (prime_p256_order - 2)))
+
+let lemma_power_step6 kInv = 
+  let a = Spec.ECDSA.exponent_spec (fromDomain_ kInv) in 
```

Comment:
for big lemmas like this, I really like `calc` statements, because they really help with proof maintenance -- it would be a good idea, for robustness of the code, to try to write calc statements if applicable

Reply:
note: I saw that almost all other places have great calc statements -- this might be coming from a previous version of the code, or might just be a particular case, but if calc didn't work for you, don't hesitate to file a bug